README.pod
=for comment Documentation generated by YARD v0.6.1 and yard-pod-plugin v0.1.

=head1 NAME

Test::Mini::Unit - Declarative Sugar for Test::Mini

=head1 DESCRIPTION

Test::Mini::Unit aims to provide a simpler, boilerplate-free environment
for writing new L<<<< S<<<<< Test::Mini >>>>>|Test::Mini >>>> test
cases.  While Test::Mini itself is a fairly reasonable environment with
very little overhead verbosity, the overhead of creating a new class --
or set of classes -- in Perl can still be a bit more distracting than
you'd really like.

=head2 Enter Test::Mini::Unit

At first glance, Test::Mini::Unit provides moderate improvements over
the traditional style, transforming this:

  package t::Test
  use base 'Test::Mini::TestCase';
  use strict;
  use warnings;

  use Test::Mini::Assertions;

  sub setup {
      # do something
  }

  sub test_some_code {
      assert($something_true);
  }

  sub teardown {
      # undo something
  }

  1;

Into this:

  use Test::Mini::Unit;

  case t::Test {
      setup {
          # do something
      }

      test some_code {
          assert($something_true);
      }

      teardown {
          # undo something
      }
  }

=head2 Advice

But Test::Mini::Unit really begins to shine as your test cases take on
more complexity.  Multiple calls to the test advice methods (C<<<< setup
>>>> and C<<<< teardown >>>>) will stack like C<<<< BEGIN >>>> and C<<<<
END >>>> blocks, allowing you to co-locate tests and relevant advice.

  # Traditional
  sub setup {
      # do a bunch of setup for test_one
      # ...
      # do a bunch of setup for test_two
      # ...
  }

  sub test_one { ... }
  sub test_two { ... }

  # Test::Mini::Unit
  setup { "do setup for test_one" }
  sub test_one { ... }

  setup { "do setup for test_two" }
  sub test_two { ... }

=head2 Test-Local Storage

Per-test local storage is automatically available as C<<<< $self >>>>
from all advice and test blocks.

   setup { $self->{data} = Package->new() }
   test data { assert_isa($self->{data}, 'Package') }

   teardown { unlink $self->{tmpfile} }

=head2 Nesting

And perhaps most usefully, test cases can be B<<<< nested >>>>.  Nested
test cases inherit all their outer scope's test advice, allowing you to
build richer tests with far less code.

  case t::IO::Scalar {
      setup { $self->{buffer} = IO::Scalar->new() }
      test can_read  { assert_can($self->{buffer}, 'read')  }
      test can_write { assert_can($self->{buffer}, 'write') }
      test is_empty  { assert_empty("@{[$self->{buffer}]}") }

      case AfterWritingString {
          setup { $self->{buffer}->print('String!') }
          test contents {
              assert_equal("@{[$self->{buffer}]}", 'String!');
          }
      }

      case AfterWritingObject {
          setup { $self->{buffer}->print($self) }
          test contents {
              assert_equal("@{[$self->{buffer}]}", "$self");
          }
      }
  }

=head2 Sharing Tests...

In some cases, you may find it useful to reuse the same tests in
different cases.  For this purpose, the C<<<< shared >>>> and C<<<<
reuse >>>> keywords exist:

  shared BasicBookTests {
      test has_pages { ... }
      test pages_have_text { ... }
  }

  case Book {
      reuse BasicBookTests;
  }

  case LargePrintBook {
      reuse BasicBookTests;
      test words_should_be_big { ... }
  }

=head2 ... And Reusing Them

Groups of shared tests may also be nested inside C<<<< case >>>> blocks,
where they will inherit the namespace of their parent.  Since shared
tests will most commonly see reuse inside either the C<<<< case >>>>
they're declared in or a nested case, it's not usually necessary to
specify the full package name. The C<<<< reuse >>>> keyword will try,
therefore, to infer the fully qualified package name from the name it's
given.  (You can always specify the full name yourself by prepending
'::'.)

  shared CommonTests {
      # __PACKAGE__ is 'Nested::CommonTests'
  }

  case Nested {
      shared CommonTests {
          # __PACKAGE__ is 'Nested::CommonTests'
      }

      case Deeply {
          shared CommonTests {
              # __PACKAGE__ is 'Nested::Deeply::CommonTests'
          }

          # includes Nested::Deeply::CommonTests
          reuse CommonTests;

          # includes Nested::CommonTests
          reuse Nested::CommonTests;

          # includes CommonTests
          reuse ::CommonTests;
      }

      # includes Nested::CommonTests
      reuse CommonTests;
  }

=head2 Automatic 'use'

To automatically use packages inside all your test cases (for example,
your own custom assertions), simply pass the 'with' option to
Test::Mini::Unit; it can accept either a single package name or an
array.

  use Test::Mini::Unit (with => [ My::Assertions, My::HelperFuncs ]);

  case t::TestCase {
      # My::Assertions and My::HelperFuncs are already imported here.
      case Nested {
          # In here, too.
      }

      shared CommonTests {
          # Yup, here too.
      }
  }




See Also:

=over

=item *

L<<<< S<<<<< Test::Mini >>>>>|Test::Mini >>>>


=back

=head1 METHODS

=head2 Class Methods

=over

=item B<<<< import >>>>

    import($class, @args)

=back

=head1 AUTHOR

Pieter van de Bruggen <pvande@cpan.org>



Changes
Revision history for Test-Mini

v1.0.2  2011-02-10
        Fixed a bug in Makefile.PL that caused failures with older versions of
        ExtUtils::MakeMaker.

        Added a proper $VERSION to the Test::Mini::Unit package.

v1.0.1  2010-10-14
        Bumped required version of Test::Mini.  Fixes test failures on Perl >=
        5.9.5.

v1.0.0  2010-09-30
        First Major Version



lib/Test/Mini/Unit.pm
# Declarative Sugar for Test::Mini.
#
# Test::Mini::Unit aims to provide a simpler, boilerplate-free environment for
# writing new {Test::Mini} test cases.  While Test::Mini itself is a fairly
# reasonable environment with very little overhead verbosity, the overhead of
# creating a new class -- or set of classes -- in Perl can still be a bit more
# distracting than you'd really like.
#
# = Enter Test::Mini::Unit
#
# At first glance, Test::Mini::Unit provides moderate improvements over the
# traditional style, transforming this:
#
#   package t::Test
#   use base 'Test::Mini::TestCase';
#   use strict;
#   use warnings;
#
#   use Test::Mini::Assertions;
#
#   sub setup {
#       # do something
#   }
#
#   sub test_some_code {
#       assert($something_true);
#   }
#
#   sub teardown {
#       # undo something
#   }
#
#   1;
#
# Into this:
#
#   use Test::Mini::Unit;
#
#   case t::Test {
#       setup {
#           # do something
#       }
#
#       test some_code {
#           assert($something_true);
#       }
#
#       teardown {
#           # undo something
#       }
#   }
#
# = Advice
#
# But Test::Mini::Unit really begins to shine as your test cases take on more
# complexity.  Multiple calls to the test advice methods (+setup+ and
# +teardown+) will stack like +BEGIN+ and +END+ blocks, allowing you to
# co-locate tests and relevant advice.
#
#   # Traditional
#   sub setup {
#       # do a bunch of setup for test_one
#       # ...
#       # do a bunch of setup for test_two
#       # ...
#   }
#
#   sub test_one { ... }
#   sub test_two { ... }
#
#   # Test::Mini::Unit
#   setup { "do setup for test_one" }
#   sub test_one { ... }
#
#   setup { "do setup for test_two" }
#   sub test_two { ... }
#
# = Test-Local Storage
#
# Per-test local storage is automatically available as +$self+ from all advice
# and test blocks.
#
#    setup { $self->{data} = Package->new() }
#    test data { assert_isa($self->{data}, 'Package') }
#
#    teardown { unlink $self->{tmpfile} }
#
# = Nesting
#
# And perhaps most usefully, test cases can be *nested*.  Nested test cases
# inherit all their outer scope's test advice, allowing you to build richer
# tests with far less code.
#
#   case t::IO::Scalar {
#       setup { $self->{buffer} = IO::Scalar->new() }
#       test can_read  { assert_can($self->{buffer}, 'read')  }
#       test can_write { assert_can($self->{buffer}, 'write') }
#       test is_empty  { assert_empty("@{[$self->{buffer}]}") }
#
#       case AfterWritingString {
#           setup { $self->{buffer}->print('String!') }
#           test contents {
#               assert_equal("@{[$self->{buffer}]}", 'String!');
#           }
#       }
#
#       case AfterWritingObject {
#           setup { $self->{buffer}->print($self) }
#           test contents {
#               assert_equal("@{[$self->{buffer}]}", "$self");
#           }
#       }
#   }
#
# = Sharing Tests...
#
# In some cases, you may find it useful to reuse the same tests in different
# cases.  For this purpose, the +shared+ and +reuse+ keywords exist:
#
#   shared BasicBookTests {
#       test has_pages { ... }
#       test pages_have_text { ... }
#   }
#
#   case Book {
#       reuse BasicBookTests;
#   }
#
#   case LargePrintBook {
#       reuse BasicBookTests;
#       test words_should_be_big { ... }
#   }
#
# = ... And Reusing Them
#
# Groups of shared tests may also be nested inside +case+ blocks, where
# they will inherit the namespace of their parent.  Since shared tests will
# most commonly see reuse inside either the +case+ they're declared in or a
# nested case, it's not usually necessary to specify the full package name.
# The +reuse+ keyword will try, therefore, to infer the fully qualified
# package name from the name it's given.  (You can always specify the full name
# yourself by prepending '::'.)
#
#   shared CommonTests {
#       # __PACKAGE__ is 'Nested::CommonTests'
#   }
#
#   case Nested {
#       shared CommonTests {
#           # __PACKAGE__ is 'Nested::CommonTests'
#       }
#
#       case Deeply {
#           shared CommonTests {
#               # __PACKAGE__ is 'Nested::Deeply::CommonTests'
#           }
#
#           # includes Nested::Deeply::CommonTests
#           reuse CommonTests;
#
#           # includes Nested::CommonTests
#           reuse Nested::CommonTests;
#
#           # includes CommonTests
#           reuse ::CommonTests;
#       }
#
#       # includes Nested::CommonTests
#       reuse CommonTests;
#   }
#
# = Automatic 'use'
#
# To automatically use packages inside all your test cases (for example, your
# own custom assertions), simply pass the 'with' option to Test::Mini::Unit;
# it can accept either a single package name or an array.
#
#   use Test::Mini::Unit (with => [ My::Assertions, My::HelperFuncs ]);
#
#   case t::TestCase {
#       # My::Assertions and My::HelperFuncs are already imported here.
#       case Nested {
#           # In here, too.
#       }
#
#       shared CommonTests {
#           # Yup, here too.
#       }
#   }
#
# @see Test::Mini
# @author Pieter van de Bruggen <pvande@cpan.org>
package Test::Mini::Unit;
use strict;
use warnings;
use 5.008;

use version 0.77; our $VERSION = qv("v1.0.2");

use Test::Mini;

require Test::Mini::Unit::Sugar::Shared;
require Test::Mini::Unit::Sugar::TestCase;

# @api private
sub import {
    my ($class, @args) = @_;
    my $caller = caller();

    strict->import;
    warnings->import;

    Test::Mini::Unit::Sugar::Shared->import(into => $caller, @args);
    Test::Mini::Unit::Sugar::TestCase->import(into => $caller, @args);
}

1;


lib/Test/Mini/Unit/Sugar/Test.pm
package Test::Mini::Unit::Sugar::Test;
use base 'Devel::Declare::Context::Simple';
use strict;
use warnings;

use Devel::Declare ();
use Sub::Name;

sub import {
    my ($class, %args) = @_;
    my $caller = $args{into} || caller;

    {
        no strict 'refs';
        *{"$caller\::test"} = sub (&) {};
    }

    Devel::Declare->setup_for(
        $caller => { test => { const => sub { $class->new()->parser(@_) } } }
    );
}

sub parser {
    my $self = shift;
    $self->init(@_);

    $self->skip_declarator;
    my $name = $self->strip_name;

    $self->inject_if_block($_) for reverse (
        $self->scope_injector_call(),
        'my $self = shift;',
    );

    $self->shadow($self->code("test_$name"));
}

sub code {
    my ($self, $name) = @_;

    my $pkg = $self->get_curstash_name;
    $name = join('::', $pkg, $name) unless ($name =~ /::/);

    return sub (&) {
        no strict 'refs';
        *{$name} = subname $name => shift;
    };
}

1;


lib/Test/Mini/Unit/Sugar/Reuse.pm
# Defines the behavior of the +reuse+ keyword.
# @api private
package Test::Mini::Unit::Sugar::Reuse;
use base 'Devel::Declare::Context::Simple';
use strict;
use warnings;

use Devel::Declare ();
use Carp qw/confess/;

sub import {
    my ($class, %args) = @_;
    my $caller = $args{into} || caller;

    {
        no strict 'refs';
        *{"$caller\::reuse"} = sub ($) {};
    }

    Devel::Declare->setup_for(
        $caller => { reuse => { const => sub { $class->new()->parser(@_) } } }
    );
}

sub parser {
    my $self = shift;
    $self->init(@_);

    $self->skip_declarator();

    my $name = $self->strip_name();
    $self->inject(qq'"$name"; ');
    my $fullname = $self->qualify_name($name);
    $self->inject("$fullname->import();");
}

sub qualify_name {
    my ($self, $name) = @_;
    my $file;
    my $mod = $self->get_curstash_name();

    if ($name =~ s/^::// || $self->get_curstash_name() eq 'main') {
        ($file = $name) =~ s/::/\//g;
        die "Cannot find module '$name' to reuse..."
            unless exists $INC{"$file.pm"};
    } else {
        my $pkg = $self->get_curstash_name();
        my @pkg_parts  = split('::', $pkg);
        my @name_parts = split('::', $name);

        unshift @pkg_parts, '';

        while (@pkg_parts) {
            $file = join('/', @pkg_parts[1..@pkg_parts-1], @name_parts);
            last if exists $INC{"$file.pm"};
            pop(@pkg_parts);
            $file = undef;
        }

        die "Cannot resolve module '$name' relative to '$pkg'..." unless $file;

        ($name = $file) =~ s/\//::/g;
    }

    return $name;
}

sub rewrite_declarator {
    my ($self, $name) = @_;

    # Parse the length of the next word.
    my $declarator = $self->declarator;
    my $length = Devel::Declare::toke_scan_word($self->offset, 0);
    confess "Couldn't find declarator '$declarator'" unless $length;

    # Verify that the next word is what we expect it to be.
    my $linestr = $self->get_linestr();
    my $found = substr($linestr, $self->offset, $length);
    confess "Expected declarator '${declarator}', got '${found}'"
        unless $found eq $declarator;

    # Replace the declarator with the given name.
    substr($linestr, $self->offset, $length) = $name;
    $self->set_linestr($linestr);
    $self->inc_offset(length($name));
}

sub inject {
  my ($self, $inject) = @_;

  my $linestr = $self->get_linestr();
  substr($linestr, $self->offset, 0) = $inject;
  $self->set_linestr($linestr);
  $self->inc_offset(length($inject));
}

1;


Makefile.PL
use strict;
use warnings;

BEGIN { require 5.008; }

use ExtUtils::MakeMaker 6.31;
use ExtUtils::Manifest  1.56;

my $SKIP_FILES = <<SKIP;
SKIP

my %WriteMakefileArgs = (
    'NAME'       => 'Test::Mini::Unit',
    'ABSTRACT'   => 'Declarative Sugar for Test::Mini',
    'VERSION'    => 'v1.0.2',
    'AUTHOR'     => 'Pieter van de Bruggen <pvande@cpan.org>',
    'LICENSE'    => 'perl',
    'META_MERGE' => {
        'resources' => {
            'homepage'   => 'http://github.com/pvande/Test-Mini-Unit',
            'repository' => 'git://github.com/pvande/Test-Mini-Unit.git',
            'bugtracker' => 'http://github.com/pvande/Test-Mini-Unit/issues',
        },
    },
    'PREREQ_PM' => {
        'version'        => '0.77',
        'Test::Mini'     => 'v1.1.2',
        'Devel::Declare' => 'v0.6.0',
        'Sub::Name'      => '0.05',
    },
    'BUILD_REQUIRES' => { },
    'CONFIGURE_REQUIRES' => {
         'ExtUtils::MakeMaker' => '6.31',
         'ExtUtils::Manifest'  => '1.56',
    },
#   'DIST_REQUIRES' => {
#       'ShipIt'                 => '0.55',
#   },
    'test' => {
        'TESTS' => 't/*.t t/*/*.t t/*/*/*.t t/*/*/*/*.t t/*/*/*/*/*.t',
    },
    'dist' => {
        'PREOP'  => 'RUBYLIB=:../yard-perl-plugin/lib:../yard-pod-plugin/lib yard -e yard-perl-plugin.rb -e yard-pod-plugin.rb',
        'POSTOP' => 'make realclean; rm MANIFEST',
    },
);

unless (eval { ExtUtils::MakeMaker->VERSION(6.56) }) {
    my $REQS = delete $WriteMakefileArgs{BUILD_REQUIRES};
    $WriteMakefileArgs{PREREQ_PM}->{$_} = $REQS->{$_} for keys %$REQS;
}

unless (eval { ExtUtils::MakeMaker->VERSION(6.52) }) {
    my $REQS = delete $WriteMakefileArgs{CONFIGURE_REQUIRES};
    $WriteMakefileArgs{PREREQ_PM}->{$_} = $REQS->{$_} for keys %$REQS;
}

(my $DISTNAME = $WriteMakefileArgs{NAME}) =~ s/::/-/g;
my $DISTDIR = "$DISTNAME-$WriteMakefileArgs{VERSION}";
my $LIBDIR = "$DISTDIR/lib";
$WriteMakefileArgs{dist}->{PREOP} .= " -o $LIBDIR; ";
$WriteMakefileArgs{dist}->{PREOP} .= "(cd $DISTDIR; find lib -name '*.pod' -type f >> MANIFEST)";

# Create a MANIFEST.SKIP file.
open(GITIGNORE, '<.gitignore');
open(SKIP, '>MANIFEST.SKIP');
print SKIP "#!include_default\n";
print SKIP "MANIFEST.SKIP.*\n";
print SKIP "^$DISTNAME.*\n";
print SKIP "^\\..*\n";
print SKIP <GITIGNORE>;
print SKIP $SKIP_FILES;
close SKIP;
close GITIGNORE;

# Create a new MANIFEST.
ExtUtils::Manifest::mkmanifest();

# Write the new Makefile.
WriteMakefile(%WriteMakefileArgs);

# Cleanup the MANIFEST.SKIP files.
unlink glob 'MANIFEST.{,SKIP}*';
